link VBS with MOSAIC and MAM5

mingxuanwupnnl opened the pull request:
This PR provides the capability of running chemUCI with MOSAIC, MAM5 strat sulfate, and VBS SOA together. This PR is a stealth feature which does not turn on VBS, MOSAIC, and MAM5 by default. 

The code modifications are mainly for:

1. couple VBS SOA gas-aerosol partition with MOSAIC, currently a temporary way. Need to work on a long term solution.
2. create new package name superfast_mam5_resus_mom_vbs_mosaic for all chem/aero 
3. fix a bug in modal_aero_deposition.F90. When calling MOSAIC, Ca and CO3 deposition are not passed to ELM. 
4. remove some developer name only comments

It is BFB (stealth) for chemUCI only simulations.
test suites results compared with NGD_v3atm baseline (./create_test $TESTNAMES --compare -b NGD_v3atm):
  PASS ERS_Ln11_P512x1.ne30pg2_EC30to60E2r2.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_rtmoff
  DIFF ERS_Ln9.ne4pg2_oQU480.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_rtmoff
  PEND PEM_Ln5.ne4pg2_oQU480.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_rtmoff
  FAIL PET_Ln5.ne4pg2_oQU480.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_rtmoff
  DIFF REP_Ln9.ne4pg2_oQU480.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm
  DIFF SMS_Ln5.ne30pg2_EC30to60E2r2.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_cosplite
  DIFF SMS_Ln5.ne4pg2_oQU480.F20TR_chemUCI-Linozv3.chrysalis_intel.eam-20tr_v3atm_cosplite

Except P3_input_dim in eam.h0 and P3_output_dim in eam.h1, other variables are BFB for tests showing DIFF.

Minimal tests passed and show BFB for custom-10_1x10, custom-10_2x5, custom-30_1x10, and custom-30_2x5.

two-yr all feature on simulation (Hui's aerosol coupling not turned on) looks okay for aerosol. Need some work on the coupling between MAM5 and MOSAIC.

hwangacme commented on components/eam/src/chemistry/modal_aero/module_mosaic_astem.F90:
```diff
@@ -245,11 +245,16 @@ subroutine ASTEM(   mcall_print_aer,
-
+#if ( defined VBS_SOA )
+  ! call ASTEM_VBS_secondary_organics(dtchem,jaerosolstate,sfc_a,Heff,phi_volatile_l,  &
+  !     integrate,aer,kg,gas,sat_soa,total_species)
```
@mingxuanwupnnl, is this line of code kept here for the coupling of SOA and MOSAIC in the future?

mingxuanwupnnl replied:
Yes. Currently, we should probably have a new subroutine on doing gas-aerosol partition between SOA aerosol and SOAG gas in MOSAIC in the future.

hwangacme commented on components/eam/src/chemistry/utils/modal_aero_deposition.F90:
```diff
@@ -473,6 +540,35 @@ subroutine set_srf_drydep(aerdepdryis, aerdepdrycw, cam_out)
       ! distributions of MAM7 fine dust and coarse dust shown in Table
       ! 1 of Liu et al (2012, doi:10.5194/gmd-5-709-2012).  In MAM3,
       ! accumulation-mode dust is assumed to resemble fine dust
+#if ( defined MOSAIC_SPECIES )
```
@mingxuanwupnnl , Please double check if the negative sign is needed for the dust dry depo terms. The wet depo and dry depo are also inconsistent in the sign convention.

mingxuanwupnnl replied:
Thank you. Just corrected them. Should not have the negative sign fro dry dep.

mingxuanwupnnl commented:
@tangq and @keziming I fixed the coupling between MOSAIC and ZMmicro (see zm_microphysics.F90). I also added a chem_mech.in file for superfast_mam5_resus_mom_vbs_mosaic. If you do not see any thing suspicious, could you re-approve it? Thank you.

tangq reviewed (approved):
@mingxuanwupnnl , I am not the right person to review the changes in MOSAIC and ZMmicro, so I will just approve, so it won't be delayed and let others to provide input.

wlin7 commented:
Merged to NGD_v3atm
